File: mobile/lib/features/library/models/read_state.dart
/// Where this profile stands in one followed series —
/// `FollowedSeriesService._read_states` in the backend, computed for a whole
/// library page in one query and sent on each list item (and on the `follow`
/// and `patch` answers, which the library swaps straight into its list).
///
/// Render it through `utils/read_state_label.dart`, never inline.
class ReadState {
  const ReadState({
    required this.started,
    this.chapterKey,
    this.chapterNumber,
    this.position,
    required this.total,
    this.latestNumber,
    this.newCount,
  });

  /// Whether this profile has opened any chapter of the series.
  final bool started;

  /// The furthest chapter opened, by reading order; null when the known list
  /// no longer carries it.
  final String? chapterKey;

  /// That chapter's printed number, where the source numbers it.
  final double? chapterNumber;

  /// 1-based position of that chapter in reading order; null when unknown.
  final int? position;

  /// How many chapters the known list holds.
  final int total;

  /// The printed number of the last chapter in reading order, where known.
  final double? latestNumber;

  /// Chapters past the furthest one opened; null when the position is unknown.
  final int? newCount;

  factory ReadState.fromJson(Map<String, dynamic> json) => ReadState(
        started: json['started'] as bool? ?? false,
        chapterKey: json['chapter_key'] as String?,
        chapterNumber: (json['chapter_number'] as num?)?.toDouble(),
        position: (json['position'] as num?)?.toInt(),
        total: (json['total'] as num?)?.toInt() ?? 0,
        latestNumber: (json['latest_number'] as num?)?.toDouble(),
        newCount: (json['new_count'] as num?)?.toInt(),
      );

  Map<String, dynamic> toJson() => {
        'started': started,
        'chapter_key': chapterKey,
        'chapter_number': chapterNumber,
        'position': position,
        'total': total,
        'latest_number': latestNumber,
        'new_count': newCount,
      };
}

File: frontend/src/features/library/read-state.ts
/**
 * What a library card says about where the reader is in a followed series.
 *
 * Every follow is created as "reading" and nobody ever changes it, so the
 * shelf used to print "Reading" under series the owner had followed as a
 * to-read note and never opened. The list now carries the server's
 * `read_state` for each row, and these helpers turn it into the card's words.
 *
 * Mirrors `mobile/lib/features/library/utils/read_state_label.dart` — both
 * clients describe the same series the same way.
 */

import { readingStatusLabel } from "./reading-stats";
import type { FollowedSeries, ReadState } from "./types";

/** Above this the pill stops counting; "3183 new" is noise, not information. */
const NEW_COUNT_CAP = 99;

function formatNumber(value: number): string {
  // 4 → "4", 12.5 → "12.5": printed chapter numbers, not measurements.
  return String(value);
}

/**
 * "Not started", "Ch 5 of 120", or null when the row carries no read state
 * (so the caller keeps whatever it said before).
 *
 * Printed numbers are used only when both ends have one and they are in
 * order; otherwise the position in the list is honest where a mix is not —
 * "Ch 12 of 10" is what a numbered chapter over an unnumbered tail would say.
 */
export function readStateLabel(state: ReadState | null | undefined): string | null {
  if (!state) return null;
  if (!state.started) return "Not started";
  const { chapter_number: number, latest_number: latest, position, total } = state;
  if (position != null) {
    if (number != null && latest != null && number <= latest) {
      return `Ch ${formatNumber(number)} of ${formatNumber(latest)}`;
    }
    return `Ch ${position} of ${total}`;
  }
  // Started, but on a chapter the known list no longer carries.
  if (number != null) return `Ch ${formatNumber(number)}`;
  return "Started";
}

/**
 * Chapters past the furthest one opened — 0 for a series never started, where
 * every chapter is "unread" but none is "new" to this reader.
 */
export function readStateNewCount(state: ReadState | null | undefined): number {
  if (!state?.started || state.new_count == null) return 0;
  return Math.max(0, state.new_count);
}

/** "2 new", "99+ new", or null for none. */
export function newCountLabel(count: number): string | null {
  if (count <= 0) return null;
  return count > NEW_COUNT_CAP ? `${NEW_COUNT_CAP}+ new` : `${count} new`;
}

/** A shelf row's note: "Ch 5 of 120 · 2 new", "Not started", or null. */
export function readStateNote(state: ReadState | null | undefined): string | null {
  const label = readStateLabel(state);
  if (label == null) return null;
  const fresh = newCountLabel(readStateNewCount(state));
  return fresh ? `${label} · ${fresh}` : label;
}

/**
 * A novel shelf row's note: where the reader is, else — for a row with no
 * read state — the reading-status word the shelf always printed.
 */
export function followedShelfNote(
  series: Pick<FollowedSeries, "reading_status" | "read_state">,
): string | null {
  const note = readStateNote(series.read_state);
  if (note) return note;
  return series.reading_status ? readingStatusLabel(series.reading_status) : null;
}

/**
 * The library grid card's and list row's meta line: where the reader is and
 * what is waiting, else — for a row with no read state — the chapter count.
 */
export function seriesCardMeta(
  series: Pick<FollowedSeries, "chapter_count" | "read_state">,
): string {
  return readStateNote(series.read_state) ?? `${series.chapter_count} chapters`;
}

/**
 * The shelf card's one muted line: where the reader is, else — for a row
 * with no read state — the chapter count it always showed, else nothing.
 */
export function followedCardSubtitle(
  series: Pick<FollowedSeries, "chapter_count" | "read_state">,
): string | null {
  const label = readStateLabel(series.read_state);
  if (label) return label;
  if (series.chapter_count <= 0) return null;
  return series.chapter_count === 1 ? "1 chapter" : `${series.chapter_count} chapters`;
}
